android/PhysicalWeb/app/src/main/java/org/physical_web/physicalweb/BroadcastActivity.java

```diff
@@ -0,0 +1,205 @@
+
+    private static final String TAG  = "BroadcastActivity";
+    private static final int ENABLE_BLUETOOTH_REQUEST_ID = 1;
+    public static final int MAX_URI_LENGTH = 18;
+    private String url;
+
+
+    @Override
+    protected void onActivityResult (int requestCode, int resultCode, Intent data) {
+        Log.d(TAG, "onActivityResult");
+        if (requestCode == ENABLE_BLUETOOTH_REQUEST_ID && resultCode == -1) {
+            checkBleAndStart();
+        } else {
+            Toast.makeText(this, getString(R.string.bt_on), Toast.LENGTH_LONG).show();
+        }
+        finish();
+    }
+
+    @Override
+    protected void onStart() {
+        super.onStart();
+        if (Build.VERSION.SDK_INT < 21){
+            Toast.makeText(this, getString(R.string.ble_os_error),
+                Toast.LENGTH_LONG).show();
+            return;
+        }
+        if (!checkIfBluetoothIsEnabled()) {
+            requestBluetooth();
+        } else {
+            checkBleAndStart();
+            finish();
+        }
+    }
+
+    private void parseAndHandleUrl(){
+        Intent intent = getIntent();
+        String action = intent.getAction();
+        String type = intent.getType();
+        String text = intent.getStringExtra(Intent.EXTRA_TEXT);
+        // Look to see if the text is a URL
+        url = findUrlInText(text);
+        // If the text is a URL
+        if (url != null) {
+            if (hasValidUrlLength(url) && isAsciiUrl(url)) {
+                // Set the url if we can
+                Log.d(TAG, "valid length");
+                startBroadcastService(url, url);
+            } else {
+                Log.d(TAG, "needs shortening");
+                // Shorten the url if necessary
+                UrlShortenerClient.ShortenUrlCallback urlSetter =
+                new UrlShortenerClient.ShortenUrlCallback() {
+                    @Override
+                    public void onUrlShortened(String newUrl) {
+                        Log.d(TAG, "shortening success");
+                        startBroadcastService(url, newUrl);
+                    }
+                    @Override
+                    public void onError(String oldUrl) {
+                        Toast.makeText(BroadcastActivity.this, getString(R.string.shorten_error),
+                            Toast.LENGTH_LONG).show();
+                    }
+                };
+                UrlShortenerClient.getInstance(this).shortenUrl(url, urlSetter, TAG);
+            }
+        } else {
+            Toast.makeText(this, getString(R.string.no_url_error),
+                Toast.LENGTH_LONG).show();
+            finish();
+        }
+    }
+
+    private void checkBleAndStart(){
+        if (checkAndWarnBleAdvertiseCapability()) {
+            parseAndHandleUrl();
+        }
+    }
+    private boolean checkAndWarnBleAdvertiseCapability() {
+        if (!hasBleAdvertiseCapability()) {
+            Toast.makeText(this, getString(R.string.ble_error),
```

Comment:
You could fold this into checkBleAndStart:

if (hasBleAdvertiseCapability())
  parseAndHandleUrl();
else
  Toast.makeText(...).show();